log/json_logger.go

```diff
@@ -17,29 +18,59 @@ func NewJSONLogger(w io.Writer) Logger {
 }
 
 func (l *jsonLogger) Log(keyvals ...interface{}) error {
-	if len(keyvals)%2 == 1 {
-		panic("odd number of keyvals")
-	}
-	m := make(map[string]interface{}, len(keyvals)/2)
+	n := (len(keyvals) + 1) / 2 // +1 to handle case when len is odd
+	m := make(map[string]interface{}, n)
 	for i := 0; i < len(keyvals); i += 2 {
-		merge(m, keyvals[i], keyvals[i+1])
+		k := keyvals[i]
+		v := ErrMissingValue
+		if i+1 < len(keyvals) {
+			v = keyvals[i+1]
+		}
+		merge(m, k, v)
 	}
 	return json.NewEncoder(l.Writer).Encode(m)
 }
 
-func merge(dst map[string]interface{}, k, v interface{}) map[string]interface{} {
+func merge(dst map[string]interface{}, k, v interface{}) {
 	var key string
 	switch x := k.(type) {
 	case string:
 		key = x
 	case fmt.Stringer:
-		key = x.String()
+		key = safeString(x)
 	default:
-		key = fmt.Sprintf("%v", x)
+		key = fmt.Sprint(x)
 	}
 	if x, ok := v.(error); ok {
-		v = x.Error()
+		v = safeError(x)
 	}
 	dst[key] = v
-	return dst
+}
+
+func safeString(str fmt.Stringer) (s string) {
```

Comment:
Good question. We only need it because of the special handling of `fmt.Stringer`. We could eliminate that case and just let `fmt.Sprint` handle them. The down side is that `fmt.Sprint` will return `"<nil>"` instead of `"NULL"`, and the angle brackets will get escaped in the JSON output, which is pretty ugly (http://play.golang.org/p/ap0GT3KO0l).